### PR TITLE
Marshalling methods with `in` parameters should have [In] attribute

### LINKED ISSUE #22641: Marshalling methods with `in` parameters should have [In] attribute
As an optimization, compiler should place `[InAttribute]` if user didn't on `in` parameters in marshaling methods. This saves an extra copy by the run-time on method return. because right now, it is treated as an ordinary `ref`, and the value is copied in and out.

cc @tannergooding @jcouv

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #22641 

cc @dotnet/roslyn-compiler @tannergooding @jcouv

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 1]
```diff
@@ -2319,6 +2319,89 @@ public class Test
...
+        {
+            var text = @"
+using System.Runtime.InteropServices;
+class T
+{
+    [DllImport(""Other.dll"")]
+    public static extern void M(in int p, int q);
+}";
+
+            CompileAndVerify(text, symbolValidator: module =>
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: Consider applying the same validator on source symbols too.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 2]
```diff
@@ -943,7 +943,18 @@ internal override bool IsMetadataOptional
             }
         }
 
-        internal sealed override bool IsMetadataIn => GetDecodedWellKnownAttributeData()?.HasInAttribute == true;
+        internal sealed override bool IsMetadataIn
+        {
+            get
+            {
+                if (RefKind == RefKind.In && ContainingSymbol.IsExtern)
```
Comment (tannergooding) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: How does `ContainingSymbol.IsExtern` impact this?

**Comment by tannergooding** [comment]
I talked with @OmarTawfik and @VSadov about this.

I don't think this should be limited to `extern` methods, as it can also impact other interop scenarios (such as COM or reverse P/Invoke).

Making this "required" is likely a breaking change, but making this optional and emitted by all future versions of the compiler is likely not.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: This attribute is only going to be emitted when applied to `extern` methods.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: > This attribute is only going to be emitted when applied to extern methods. 

I see no reason for making this restriction.

---
In reply to: [149226708](https://github.com/dotnet/roslyn/pull/23031#discussion_r149226708) [](ancestors = 149226708)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 3]
```diff
@@ -2319,6 +2319,89 @@ public class Test
...
+            {
+                var parameters = module.GlobalNamespace.GetTypeMember("T").GetMethod("M").GetParameters();
+                Assert.Equal(2, parameters.Length);
+
+                Assert.False(parameters[0].IsMetadataIn);
+                Assert.False(parameters[1].IsMetadataIn);
+            });
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: Other tests to consider:

- When `in` is used in a delegate signature
- When a parameter has the signature `[Out] in int e`

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceSimpleParameterSymbol.cs [hunk 4]
```diff
@@ -64,7 +64,7 @@ internal override bool IsExtensionMethodThis
 
         internal override bool IsMetadataIn
         {
-            get { return false; }
+            get { return RefKind == RefKind.In && ContainingSymbol.IsExtern; }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceSimpleParameterSymbol.cs]: > && ContainingSymbol.IsExtern [](start = 46, length = 29)

I see no reason for making this restriction.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceSimpleParameterSymbol.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ internal override bool IsExtensionMethodThis
 
         internal override bool IsMetadataIn
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceSimpleParameterSymbol.cs]: >IsMetadataIn [](start = 31, length = 12)

It feels like we should adjust implementation of this property for SynthesizedParameterSymbolBase.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs (hunk 3 above)
Comment (tannergooding) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: @jaredpar, the second was made illegal for 7.2

Comment (tannergooding) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: https://github.com/dotnet/roslyn/pull/22645

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1).

**Comment by OmarTawfik** [comment]
per @tannergooding's comment. I'll block this PR until a decision from LDM is taken on the scope of the change (extern vs everywhere).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: There is a concrete reason to add it to extern methods: affects marshaling semantics. Doing it everywhere opens up more scenarios for instance:

- When I "implement interface" in VB for a C# method that had `in` will it add the `<In>` attribute to the generated source? That doesn't seem to add any value. Either way need to make a call here (and write tests)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs [hunk 6]
```diff
@@ -2318,7 +2318,7 @@ public class Test
                 //     public static int operator + (in Test x, in Test y) => 0;
                 Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "in Test y").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute", ".ctor").WithLocation(11, 46));
         }
-
+
```
Comment (tannergooding) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs]: nit: this line contains whitespace

**Comment by tannergooding** [comment]
Should you add a NoPIA test?

**Comment by OmarTawfik** [comment]
@jaredpar per tests added, In bit is copied to proxy methods if they have `in` parameters, but user placed `[In]` attributes are not copied.

File: src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs [hunk 7]
```diff
@@ -2485,5 +2485,415 @@ public void PEHeaders2()
...
+                sourceSymbolValidator: module =>
+            {
+                var parameters = module.ContainingAssembly.GetTypeByMetadataName("D").DelegateInvokeMethod.Parameters;
+                Assert.Equal(4, parameters.Length);
+
+                Assert.True(parameters[0].IsMetadataIn);
+                Assert.True(parameters[1].IsMetadataIn);
+                Assert.True(parameters[2].IsMetadataIn);
+                Assert.False(parameters[3].IsMetadataIn);
+            },
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EmitMetadataTests.cs]: It looks like the offset is off.

**Comment by OmarTawfik** [comment]
@jaredpar any other comments?

**Pull request merged**